### PR TITLE
Support match, match aggregate, match group, and match group aggregate

### PR DESCRIPTION
**Pull request by lolski**
## What is the goal of this PR?

We have updated protocol to support not only `match` queries, but also `match aggregate`, `match group`, and `match group aggregate`.

## What are the changes implemented in this PR?

- Extend the protocol by adding new message types: `Match.Aggregate`, `Match.Group`, and `Match.Group.Aggregate`
- Remove the ability to nest `AnswerGroup` which is no longer needed given the new answer structure.

### CODE REVIEW

File: protobuf/query.proto [hunk 1]
```diff
@@ -58,6 +58,36 @@ message Graql {
...
+            }
+        }
+
+        message Group {
+            message Req {
+                string query = 1;
+            }
+
+            message Res {
+                repeated AnswerGroup answers = 1;
```
Comment (lolski) [inline protobuf/query.proto:78]: Should we be even more type-safe and make `AnswerGroup` more specific: `AnswerGroup` and `AnswerGroup.Aggregate`?

File: protobuf/answer.proto [hunk 2]
```diff
@@ -26,9 +26,8 @@ package grakn.protocol;
 
 message Answer {
     oneof answer {
-        AnswerGroup answer_group = 1;
-        ConceptMap concept_map = 2;
-        Number number = 3;
+        ConceptMap concept_map = 1;
+        Number number = 2;
```
Comment (alexjpwalker) [inline protobuf/answer.proto:30]: I wonder if we should pick a less ambiguous term here - `NumericAnswer numeric = 2`, perhaps?

File: protobuf/query.proto [hunk 3]
```diff
@@ -58,6 +58,36 @@ message Graql {
...
             repeated ConceptMap answers = 1;
         }
+
+        message Aggregate {
+            message Req {
+                string query = 1;
+            }
+
+            message Res {
+                Number answers = 1;
```
Comment (alexjpwalker) [inline protobuf/query.proto:68]:
```suggestion
                Number answer = 1;
```

File: protobuf/query.proto (hunk 1 above)
Comment (alexjpwalker) [inline protobuf/query.proto:78]:
```suggestion
                repeated AnswerGroup answer_groups = 1;
```

File: protobuf/query.proto [hunk 4]
```diff
@@ -58,6 +58,36 @@ message Graql {
...
+                repeated AnswerGroup answers = 1;
+            }
+
+            message Aggregate {
+                message Req {
+                    string query = 1;
+                }
+
+                message Res {
+                    repeated AnswerGroup answers = 1;
```
Comment (alexjpwalker) [inline protobuf/query.proto:87]:
```suggestion
                    repeated AnswerGroup answer_groups = 1;
```

File: protobuf/query.proto (hunk 1 above)
Comment (alexjpwalker) [reply protobuf/query.proto:78]: I think that makes sense, yes - it also follows the structure of the Java Graql repo. `GraqlMatch`, `GraqlMatch.Group`, `GraqlMatch.Aggregate`, `GraqlMatch.Group.Aggregate`

**Review by alexjpwalker — CHANGES_REQUESTED** [review]
Made some suggestions.

File: protobuf/answer.proto (hunk 2 above)
Comment (alexjpwalker) [reply protobuf/answer.proto:30]: We decided not to.

**Comment by alexjpwalker** [comment]
We decided to:

- move `Match`, `Insert`, `Delete` etc. inside `Query` (instead of where it currently lives, inside `Graql`) and delete `Graql`.
- delete `Answer`
- rename `AnswerGroup` to `ConceptMapGroup`
- create `NumberGroup`
- name the new query type messages: `MatchGroup`, `MatchAggregate`, `MatchGroupAggregate`

@lolski

File: protobuf/answer.proto [hunk 5]
```diff
@@ -24,25 +24,22 @@ import "protobuf/concept.proto";
...
-}
-
 message ConceptMap {
     map<string, Concept> map = 1;
     string pattern = 2;
 }
 
+message ConceptMapGroup {
+    Concept owner = 1;
+    repeated ConceptMap concept_map = 2;
```
Comment (alexjpwalker) [inline protobuf/answer.proto:34]: Let's make sure all our `repeated` fields have plural names- this is particularly beneficial to client Python, which uses the field names themselves as property names:
```py
answers: list = match_res.answers
```
is more readable than
```py
answers: list = match_res.answer
```
```suggestion
    repeated ConceptMap concept_maps = 2;
```

File: protobuf/answer.proto [hunk 6]
```diff
@@ -24,25 +24,22 @@ import "protobuf/concept.proto";
...
+}
+
 message Number {
     string value = 1;
     // We use a string to contain the full representation of a number
 }
+
+message NumberGroup {
+    Concept owner = 1;
+    repeated Number number = 2;
```
Comment (alexjpwalker) [inline protobuf/answer.proto:44]: See comment above

File: protobuf/query.proto [hunk 7]
```diff
@@ -30,26 +30,23 @@ message Query {
...
-            Graql.Delete.Req delete_req = 100;
-            Graql.Define.Req define_req = 101;
-            Graql.Undefine.Req undefine_req = 102;
-            Graql.Match.Req match_req = 103;
-            Graql.Insert.Req insert_req = 104;
+            Delete.Req delete_req = 100;
+            Define.Req define_req = 101;
+            Undefine.Req undefine_req = 102;
+            Match.Req match_req = 103;
+            Insert.Req insert_req = 104;
```
Comment (alexjpwalker) [inline protobuf/query.proto:37]: We additionally need:
```protobuf
MatchAggregate.Req match_aggregate_req = 105;
MatchGroup.Req match_group_req = 106;
MatchGroupAggregate.Req match_group_aggregate_req = 107;
```
And the matching responses should be added to `Res`, below

**Review by alexjpwalker — CHANGES_REQUESTED** [review]

File: protobuf/answer.proto (hunk 6 above)
Comment (lolski) [reply protobuf/answer.proto]: Fixed in https://github.com/graknlabs/protocol/pull/100/commits/8e16eed30e9a1221201d3b4ff85108dd6874ea86

File: protobuf/answer.proto (hunk 5 above)
Comment (lolski) [reply protobuf/answer.proto]: Fixed in https://github.com/graknlabs/protocol/pull/100/commits/8e16eed30e9a1221201d3b4ff85108dd6874ea86

File: protobuf/query.proto (hunk 7 above)
Comment (lolski) [reply protobuf/query.proto]: Fixed in https://github.com/graknlabs/protocol/pull/100/commits/e6354fe87b04c8f37c5c81c8387d24c5fdcedfa4

**Review by alexjpwalker — APPROVED** [review]

**Pull request merged**